File: stremio-addon/src/tmdb.js
import fs from "node:fs";
import path from "node:path";

const CACHE_FILE = path.resolve("tmdb_cache.json");
const TMDB_BASE = "https://api.themoviedb.org/3";
const API_KEY = process.env.TMDB_API_KEY;

let cache = {};
let cacheLoaded = false;

function loadCache() {
  if (cacheLoaded) return;
  try {
    cache = JSON.parse(fs.readFileSync(CACHE_FILE, "utf-8"));
  } catch {
    cache = {};
  }
  cacheLoaded = true;
}

function saveCache() {
  fs.writeFileSync(CACHE_FILE, JSON.stringify(cache, null, 2));
}

function cacheKey(title, year, type) {
  return `${title.toLowerCase().trim()}|${year}|${type || ""}`;
}

function delay(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

let lastRequest = 0;
const RATE_LIMIT_MS = 300; // ~3 req/s, well under TMDB's 40/10s

async function tmdbRequest(endpoint) {
  if (!API_KEY) return null;
  const now = Date.now();
  const wait = RATE_LIMIT_MS - (now - lastRequest);
  if (wait > 0) await delay(wait);
  lastRequest = Date.now();

  const url = `${TMDB_BASE}${endpoint}${endpoint.includes("?") ? "&" : "?"}api_key=${API_KEY}`;
  try {
    const res = await fetch(url);
    if (!res.ok) return null;
    return res.json();
  } catch {
    return null;
  }
}

async function fetchImdbId(title, year, expectedType) {
  const params = new URLSearchParams({ query: title });
  if (year) params.set("year", year);
  const data = await tmdbRequest(`/search/multi?${params}`);
  if (!data?.results?.length) return null;

  let best;
  if (expectedType) {
    const targetMediaType = expectedType === "series" ? "tv" : "movie";
    best = data.results.find((r) => {
      const mt = r.media_type || (r.title ? "movie" : "tv");
      return mt === targetMediaType;
    });
    if (!best) best = data.results[0];
  } else {
    best = data.results[0];
  }

  const tmdbId = best.id;
  const mediaType = best.media_type || (best.title ? "movie" : "tv");
  const type = mediaType === "movie" ? "movie" : "tv";

  const ext = await tmdbRequest(`/${mediaType}/${tmdbId}/external_ids`);
  const imdbId = ext?.["imdb_id"];

  return {
    imdbId: imdbId || null,
    tmdbId: String(tmdbId),
    type,
    poster: best.poster_path ? `https://image.tmdb.org/t/p/w500${best.poster_path}` : null,
    description: best.overview || null,
    genres: best.genre_ids ? [] : null,
  };
}

export async function getTmdbMeta(title, year, expectedType) {
  loadCache();
  const key = cacheKey(title, year, expectedType);

  if (cache[key]) return cache[key];

  const result = await fetchImdbId(title, year, expectedType);
  if (result) {
    cache[key] = result;
    saveCache();
  } else {
    cache[key] = null;
    saveCache();
  }
  return result;
}

export function getCachedMeta(title, year) {
  loadCache();
  return cache[cacheKey(title, year)] || null;
}

File: fastapi-addon/tmdb.py
from __future__ import annotations

from typing import Any

import httpx


TMDB_BASE = "https://api.themoviedb.org/3"
_cache: dict[str, dict[str, Any]] = {}


def _cache_key(title: str, year: str, content_type: str) -> str:
    return f"{title.lower().strip()}|{year}|{content_type}"


async def _tmdb_get(endpoint: str, api_key: str) -> dict[str, Any] | None:
    if not api_key:
        return None

    separator = "&" if "?" in endpoint else "?"
    url = f"{TMDB_BASE}{endpoint}{separator}api_key={api_key}"

    try:
        async with httpx.AsyncClient(timeout=20) as client:
            response = await client.get(url)
    except httpx.RequestError:
        return None

    if response.status_code >= 400:
        return None

    return response.json()


async def get_tmdb_meta(
    title: str,
    year: str = "",
    expected_type: str = "",
    api_key: str = "",
) -> dict[str, Any] | None:
    key = _cache_key(title, year, expected_type)
    if key in _cache:
        return _cache[key]

    if not api_key:
        return None

    params = httpx.QueryParams({"query": title, **({"year": year} if year else {})})
    data = await _tmdb_get(f"/search/multi?{params}", api_key)
    if not data or not data.get("results"):
        return None

    target = "tv" if expected_type == "series" else "movie"
    best = next(
        (
            result
            for result in data["results"]
            if (result.get("media_type") or ("movie" if result.get("title") else "tv")) == target
        ),
        data["results"][0],
    )

    media_type = best.get("media_type") or ("movie" if best.get("title") else "tv")
    tmdb_type = "series" if media_type == "tv" else "movie"
    external_ids = await _tmdb_get(f"/{media_type}/{best['id']}/external_ids", api_key)

    result = {
        "imdb_id": (external_ids or {}).get("imdb_id"),
        "tmdb_id": str(best["id"]),
        "type": tmdb_type,
        "poster": f"https://image.tmdb.org/t/p/w500{best['poster_path']}" if best.get("poster_path") else None,
        "description": best.get("overview"),
        "genres": [],
    }
    _cache[key] = result
    return result
